src/kubernetes_operator/iam_mapping.py

```diff
@@ -0,0 +1,221 @@
+    identities = identities if isinstance(identities, list) else list()
+
+    for identity_mapping in identity_mappings["items"]:
+        identities = ensure_identity(identity_mapping["spec"], identities)
+
+    asyncio.run(apply_identity_mappings(configmap, identities))
+
+
+def get_identity_mappings(configmap: V1ConfigMap) -> list:
+    """Get the identity mappings from the aws-auth configmap as a list.
+
+    :return identities: The combined user and role mapping list
+    """
+    try:
+        identities = []
+        identities.extend(yaml.safe_load(configmap.data["mapUsers"]))
+        identities.extend(yaml.safe_load(configmap.data["mapRoles"]))
+
+        return identities
+    except yaml.YAMLError as yaml_error:
+        logger.warning("Error loading configmap mappings: %s", yaml_error)
+        return []
+
+
+async def apply_identity_mappings(existing_cm: V1ConfigMap, identity_mappings: list) -> None:
+    """Apply new identity mappings to override the existing aws-auth mapping.
+
+    :param existing_cm: The current configmap
+    :param identity_mappings: The new identity mappings
+    """
+    user_mappings = []
+    role_mappings = []
+    for identity_mapping in identity_mappings:
+        if identity_mapping.get("userarn") is not None:
+            user_mappings.append(identity_mapping)
+        elif identity_mapping.get("rolearn") is not None:
+            role_mappings.append(identity_mapping)
+        else:
+            logger.warning("Unrecognized mapping. Skipping %s", identity_mapping)
+
+    existing_cm.data["mapUsers"] = yaml.safe_dump(user_mappings)
+    existing_cm.data["mapRoles"] = yaml.safe_dump(role_mappings)
+    API.patch_namespaced_config_map("aws-auth", "kube-system", existing_cm)
+
+
+def ensure_identity(identity: dict, identity_list: list) -> list:
+    """Ensure the identity is in the list and update it if it is, add the identity if not present.
+
+    :param identity: The identity to check
+    :param identity_list: The list to check against
+    :return list: The updated list
+    """
+
+    for i, existing_identity in enumerate(identity_list):
+        # Handle existing identity
+        if existing_identity["username"] == identity["username"]:
+            identity_list[i] = identity
+            return identity_list
+    # Handle new identity
+    identity_list.append(identity)
+    return identity_list
+
+
+def delete_identity(identity: dict, identity_list: list) -> list:
+    """Delete an identity from the identity list if present.
+
+    :param identity: The identity to delete
+    :param identity_list: The list of identities
+    :return list: The updated list
+    """
+
+    for i, existing_user in enumerate(identity_list):
+        if existing_user["username"] == identity["username"]:
+            del identity_list[i]
+            return identity_list
+
+    logger.warning("Failed to delete %s, identity was not found", identity["username"])
+    return identity_list
```

Comment:
Those 2 functions would probably have triggered me to rewrite this using Dataclasses and Sets instead of manually iterating over the list and updating it. But hey who Am I to judge ? :stuck_out_tongue:

Reply:
Definitely agree, although I did say we could improve it in this PR, it wasn't the main goal. So I am thinking of taking a pass on this one. There aren't any tests so if I rewrite I would also have to test. I think I'll leave it to you guys later on if you need it ;)